sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+
+  read_only_key(bool readOnly);
+};
+
+struct zero_init_key {
+  using value_t = property_value<zero_init_key>;
+
+  read_only_key(bool zeroInit);
+};
+
+struct use_existing_memory_key {
+  using value_t = property_value<use_existing_memory_key>;
+
+  use_existing_memory(void *ptr, size_t size);
+};
+
+inline constexpr initial_threshold_key::value_t initial_threshold;
+inline constexpr maximum_size_key::value_t maximum_size;
+inline constexpr read_only_key::value_t read_only;
+inline constexpr zero_init_key::value_t zero_init;
+inline constexpr use_existing_memory_key::value_t use_existing_memory;
+
+}  // ext::oneapi::experimental
+----
+
+|===
+|Property|Description
+
+|`initial_threshold`
+|The `initial_threshold` property specifies the initial deallocation threshold
+ value for the memory pool. If this property is not used the default value is
+ zero, and this can be increased after the memory pool is created by calling
+ `memory_pool::set_new_threshold`.
+
+|`maximum_size`
+|The `maximum_size` property specifies the maximum size of the memory pool,
+ after which any allocation will result in an exception. If the value specified
+ is larger than the implementation can support an exception with the
+ `errc::memory_allocation` error code is thrown. If this property is not used
+ the default value is implementation-defined.
+
+|`read_only`
+|The `read_only` property is a performance hint which asserts that all memory
+ allocations from the memory pool will only ever be read from, this can be used
+ by the SYCL runtime to optimize for performance.
+
+|`zero_init`
+|The `zero_init` property adds the requirement that all memory allocated to the
+ memory as it is allocated from the source to the memory pool will be
+ initialised to zero. Note there is no guarantee that the memory allocation be
+ re-initialized to zero when it is re-allocated from the pool, so users must
+ re-initialize memory to zero if they wish for later allocations to have this
+ behavior.
+
+|`use_existing_memory`
+|The `use_existing_memory` property adds the requirement that the memory pool
+ will use an existing USM memory allocation provided by the user instead of
+ allocating from the OS. This property takes a pointer to a valid USM memory
+ allocation of the same allocation kind as the memory pool is initialized with
+ and the size of that memory allocation. Using this property will implicitly set
+ the `maximum_size` and `initial_threshold` property to be that of the size
+ provided, and as such using the `maximum_size` or `initial_threshold`
+ properties in conjunction with this property will cause the `memory_pool`
+ constructor to throw an exception with the `errc::invalid` error code.
+
+|===
+
+
+=== Default memory pools
+
+As well as being able to construct a memory pool explicitly, this extension
+introduces a default memory pool per device for each SYCL context and device
+pair for device allocations and a default memory pool per context for host
+allocations.
```

Comment:
There is also USM "system" memory.  Is there a default memory pool for that?

Reply:
There's currently not default memory pool for shared or system allocations, we could add this if there is a desire for it though.

Reply:
What was the rationale for adding default pools for "device" and "host" but not "shared"?

Reply:
The rationale for this was simply that the motivational use cases were for host or device allocations and not shared. We considered adding default memory pools for shared allocations as well, we added an open issue for it at the end of the extension. After thinking about this some more, there's no reason not to include this as well, so I've altered the interface to allow for shared allocations as well in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

Reply:
On the question of system USM, we could potentially add this, it could be useful to provide a way to allocate system USM asynchronously, though it may not be so straight forward as system USM uses the system malloc/free so there's no `usm::alloc` enumeration for system USM.

Reply:
Yes, I agree that it makes sense to omit "system" USM for now.